### ISSUE TITLE
MueLu/Amesos2/Tpetra: new Failures in empire

### ISSUE THREAD

**Issue by rppawlo** [trilinos/Trilinos#6264]
## Bug Report
@trilinos/amesos2 
Starting last night, empire tests are failing in builds against trilinos develop.

The hanging tests are reporting:
```
Set 1 -> Set 1 -> Set 1 -> Set 1 -> 
...

*********** Caught Exception: Begin Error Report ***********
/scratch/mbetten/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:277:

Throw number = 1

Throw test that evaluated to true: info > 0

KLU2 numeric factorization failed
 ```

The stack trace:
```
    at ../../.././libstdc++-v3/libsupc++/eh_throw.cc:75
#1  0x00007fffcf0b9b8c in Amesos2::KLU2<Tpetra::CrsMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >, Tpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > >::numericFactorization_impl (
    this=0x5ba82b0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:276
#2  0x00007fffcf12a43d in Amesos2::SolverCore<Amesos2::KLU2, Tpetra::CrsMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >, Tpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > >::numericFactorization (this=0x5ba82b0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp:153
#3  0x00007fffcf12ab29 in Amesos2::SolverCore<Amesos2::KLU2, Tpetra::CrsMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >, Tpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > >::solve (
    this=0x5ba82b0, X=..., B=...)
---Type <return> to continue, or q <return> to quit---
    at /scratch/mbetten/Trilinos/Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp:208
#4  0x00007fffcf12a505 in Amesos2::SolverCore<Amesos2::KLU2, Tpetra::CrsMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >, Tpetra::MultiVector<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> > >::solve (
    this=0x5ba82b0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp:165
#5  0x00007fffdefc6074 in MueLu::Amesos2Smoother<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Apply (
    this=0x5c06110, X=..., B=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/muelu/src/Smoothers/MueLu_Amesos2Smoother_def.hpp:316
#6  0x00007fffdf188d41 in MueLu::DirectSolver<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Apply (
    this=0x5a528d0, X=..., B=..., InitialGuessIsZero=true)
    at /scratch/mbetten/Trilinos/Trilinos/packages/muelu/src/Smoothers/MueLu_DirectSolver_def.hpp:188
#7  0x00007fffdf24530a in MueLu::Hierarchy<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::Iterate (
    this=0x5bb2690, B=..., X=..., conv=..., InitialGuessIsZero=true, 
    startLevel=0)
---Type <return> to continue, or q <return> to quit---
    at /scratch/mbetten/Trilinos/Trilinos/packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:977
#8  0x00007fffe0dc2c5e in MueLu::RefMaxwell<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::applyInverseAdditive (this=0x5a01730, RHS=..., X=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:1864

#9  0x00007fffe0dbfdb2 in MueLu::RefMaxwell<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::apply (
    this=0x5a01730, RHS=..., X=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp:2102
#10 0x00007fffe6fc64a4 in Thyra::XpetraLinearOp<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::applyImpl

    (this=0x5c0e260, M_trans=Thyra::NOTRANS, X_in=..., Y_inout=..., alpha=1, 
    beta=0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/muelu/adapters/stratimikos/Thyra_XpetraLinearOp_def.hpp:173
#11 0x00007fffedcf8fc8 in Thyra::LinearOpBase<double>::apply (this=0x5c0e260, 
    M_trans=Thyra::NOTRANS, X=..., Y=..., alpha=1, beta=0)
    at /scratch/mbetten/installs/install-for-PANZER/GNU-7.2.0-DEBUG-OPENMP/include/Thyra_LinearOpBase_decl.hpp:300
#12 0x00007fffc955a7cf in Thyra::apply<double> (M=..., M_trans=Thyra::NOTRANS, 
---Type <return> to continue, or q <return> to quit---
    X=..., Y=..., alpha=1, beta=0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/thyra/core/src/interfaces/operator_vector/fundamental/Thyra_LinearOpBase_def.hpp:82
#13 0x00007fffd809de13 in Belos::OperatorTraits<double, Thyra::MultiVectorBase<double>, Thyra::LinearOpBase<double> >::Apply (Op=..., x=..., y=..., 
    trans=Belos::NOTRANS)
    at /scratch/mbetten/Trilinos/Trilinos/packages/stratimikos/adapters/belos/src/BelosThyraAdapter.hpp:606
#14 0x00007fffd81cee9b in Belos::LinearProblem<double, Thyra::MultiVectorBase<double>, Thyra::LinearOpBase<double> >::applyRightPrec (this=0x5b4b590, x=..., 
    y=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/belos/src/BelosLinearProblem.hpp:1108
#15 0x00007fffd8224027 in Belos::CGIter<double, Thyra::MultiVectorBase<double>, Thyra::LinearOpBase<double> >::initializeCG (this=0x5979b60, newstate=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/belos/src/BelosCGIter.hpp:406
#16 0x00007fffd817f8f2 in Belos::PseudoBlockCGSolMgr<double, Thyra::MultiVectorBase<double>, Thyra::LinearOpBase<double>, true>::solve (this=0x5a5d7f0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/belos/src/BelosPseudoBlockCGSolMgr.hpp:822
#17 0x00007fffd808fbe9 in Thyra::BelosLinearOpWithSolve<double>::solveImpl (
    this=0x5862b10, M_trans=Thyra::NOTRANS, B=..., X=..., solveCriteria=...)
---Type <return> to continue, or q <return> to quit---
    at /scratch/mbetten/Trilinos/Trilinos/packages/stratimikos/adapters/belos/src/Thyra_BelosLinearOpWithSolve_def.hpp:657
#18 0x00007ffff6f9faa5 in Thyra::LinearOpWithSolveBase<double>::solve (
    this=0x5862b10, A_trans=Thyra::NOTRANS, B=..., X=..., solveCriteria=...)
    at /scratch/mbetten/installs/install-for-PANZER/GNU-7.2.0-DEBUG-OPENMP/include/Thyra_LinearOpWithSolveBase_decl.hpp:456
#19 0x00007fffedcc10b3 in empire::ElectroMagneticSolverInterface::applyJacobianInverse (this=0x132b8c0, increment=..., rhs=...)
    at ../src/em_solvers/ElectroMagneticSolverInterface.cpp:854
#20 0x00000000006aa135 in empire::LinearSolverWrapper::solveImpl (
    this=0x599fc10, residual=..., solution=...)
    at ../src/em_solvers/LinearSolverWrapper.hpp:77
#21 0x00007ffff6f9faa5 in Thyra::LinearOpWithSolveBase<double>::solve (
    this=0x599fc10, A_trans=Thyra::NOTRANS, B=..., X=..., solveCriteria=...)
    at /scratch/mbetten/installs/install-for-PANZER/GNU-7.2.0-DEBUG-OPENMP/include/Thyra_LinearOpWithSolveBase_decl.hpp:456

#22 0x00007fffe6e8c78c in Thyra::solve<double> (A=..., A_trans=Thyra::NOTRANS, 
    B=..., X=..., solveCriteria=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/thyra/core/src/interfaces/operator_solve/fundamental/Thyra_LinearOpWithSolveBase_decl.hpp:561
#23 0x00007fffdc8f24be in NOX::Thyra::Group::applyJacobianInverseMultiVector (
    this=0x598f970, p=..., input=..., result=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src-thyra/NOX_Thyra_Group---Type <return> to continue, or q <return> to quit---
.C:807
#24 0x00007fffdc8f16d9 in NOX::Thyra::Group::applyJacobianInverse (
    this=0x598f970, p=..., input=..., result=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src-thyra/NOX_Thyra_Group.C:665
#25 0x00007fffdc8f164a in NOX::Thyra::Group::applyJacobianInverse (
    this=0x598f970, p=..., input=..., result=...)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src-thyra/NOX_Thyra_Group.C:656
#26 0x00007fffdc87177a in NOX::Solver::SingleStep::try_step (this=0x5693ba0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src/NOX_Solver_SingleStep.C:179
#27 0x00007fffdc871c0e in NOX::Solver::SingleStep::step (this=0x5693ba0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src/NOX_Solver_SingleStep.C:209
#28 0x00007fffdc871d1c in NOX::Solver::SingleStep::solve (this=0x5693ba0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src/NOX_Solver_SingleStep.C:227
#29 0x00007fffdc9183f5 in Thyra::NOXNonlinearSolver::solve (this=0x598e680, 
    x=0x56de670, delta=0x0)
    at /scratch/mbetten/Trilinos/Trilinos/packages/nox/src-thyra/Thyra_NonlinearSolver_NOX.cpp:241
#30 0x00007fffe22dcae9 in Tempus::StepperImplicit<double>::solveImplicitODE (

```

### Steps to Reproduce
any build of empire on atdm supported platforms fails.

**Comment by srajama1** [comment]
@rppawlo Amesos2 and KLU2 hasn't changed in months. My suspicion is something has changed in matrix construction that ends up giving a singular matrix to Amesos2. This exception was put in place specifically for that purpose during a similar bug few months ago.

**Comment by rppawlo** [comment]
@srajama1 - empire passes with their fork of trilinos that updated cleanly from trilinos develop two days ago. A change to trilinos from a merge yesterday is triggering this failure.

**Comment by rppawlo** [comment]
Here's a list of the commits not in empire fork. I'll start a bisection and try to find the issue. 

```
breve > git log --oneline em-plasma/develop..github/develop
a3bd08f Merge pull request #6190 from brian-kelley/FixAdd
0156125 Merge Pull Request #6257 from bartlettroscoe/Trilinos/atdv-228-fix-sems-rhel7
5188171 ATDM: Fix detection of sems-rhel6 and sems-rhel7 when SEMS_SYSTEM is not set (ATDV-228)
618dc53 Merge Pull Request #6232 from trilinos/Trilinos/tempus-add-ssp-rk-methods
d475fc6 Merge Pull Request #6252 from bartlettroscoe/Trilinos/atdv-228-fix-cdash-hostname
5cc83da Merge Pull Request #6248 from CamelliaDPG/Trilinos/fix-6246
35330d7 Merge Pull Request #6196 from lucbv/Trilinos/MueLu_region_unit_tests
7d70506 Intrepid2: use standard idiom for argument-dependent lookup, per @mhoemmen.
7bcfbf9 Merge Pull Request #6235 from cgcgcg/Trilinos/refmaxwellPR
c2f14a2 Tpetra: fix syntax error from last 6190 commit
6004ea7 Tempus: update ExplicitRK convergence test
4e48a84 Tempus: update DIRK convergence test
172ebf7 Temmpus: add SSPDIRK33 - stage = 3, order = 3 TVD scheme
8b238db Tempus: add SSPDIRK23 - stage = 2, order = 3 TVD method
7aaebd8 Tempus: add SSPDIRK32 - stage = 3, order = 2 TVD method
c48c62e Tempus: add SSPDIRK22 - stage = 2, order = 2 TVD method
6d18b4e Tempus: add SSPERK54 methods
3a899ed Tempus: 'RK Explicit 3 Stage 3rd order TVD' alias 'SSPERK33'
d61c93c Tempus: 'RK Explicit Trapezoidal' alias 'SSERK22'
1bbbbd5 MueLu: adding region matrix unit-test
e430431 Merge Pull Request #6253 from cgcgcg/Trilinos/mueluFusedPR
e76f1cf Merge Pull Request #6250 from bartlettroscoe/Trilinos/git-email-map-update
acc8bcf Merge Pull Request #6251 from trilinos/Trilinos/tribits_github_snapshot
d631c9f MueLu ParameterListInterpreter: fuse prolongation and update
c2c3e24 Fix setting of ATDM_CONFIG_BUILD_NAME when hostname match system is in build name (ATDV-228)
75c8ae7 Tpetra: fix spelling, use teuchos test/throw
69abace Automatic snapshot commit from tribits at 83aecff
9c75233 Update mapping of git users to remove duplicate and put in sandia.gov addresses
db3f191 Intrepid2: add "using getValues" declaration to avoid hidden virtual function warnings.
196e309 Intrepid2: adjusting floating point comparisons in tests to be more robust to platform differences.
8a3cf79 Tpetra: pass parallel-for index by value
643c538 Tpetra: Final cleanup for #6190
0cb3d27 MueLu RefMaxwell: Use same extensions as usual data export
a89c978 MueLu export data: Label files with hierarchy object label
3e45348 MueLu RefMaxwell: row sum dropping
5d16cca MueLu RefMaxwell: Fix "implicitTranspose" code path
1851d87 MueLu RefMaxwell: Allow smoothing of special prolongator
735d02a Tpetra/Xpetra Misc cleanup
762d55a Xpetra: fixed TwoMatrixAdd result being fillComplete
56f32a4 Tpetra: cleanup unused comm
adf5296 Xpetra: TwoMatrixAdd() has codepath for !fc A,B
eda4c94 MueLu RefMaxwell: Change splitting into subcommunicators
a8972eb MueLu RefMaxwell: Improve output
cf0971b Xpetra: revert leftover whitespace changes
b8063c4 Tpetra: remote map dump from debuggin
2277573 Tpetra: added makeColMap() that takes View<GO>
95ad008 Tpetra: set packed, 1D, global colinds in CrsGraph
b2ec041 Tpetra: MM::add() col map has locals before remotes
4decca1 Xpetra: fix other two versions of TwoMatrixAdd()
a17cd2a Tpetra: fixed index base of sum col map
2333878 Tpetra/Xpetra: fixes for add()
70a2866 Tpetra: use RCP<const Map> instead of RCP<Map>
d9a226a Tpetra: fixed mm add col map index base (should be 0)
ad9d794 Tpetra MM add(): handle zero local rows case safely
fb656a4 Cleanup and fixes for #6190
75333e4 Tpetra: fix add() column map
```

**Comment by rppawlo** [comment]
looks like the culprit is #6190 . reverting this merge fixes the problem. A test problem in empire to reproduce is:

DualFeedEMInlineMesh-mpi-parallel

**Comment by rppawlo** [comment]
Another test in core em solver (no need for pic): SourceCurrentQuiescentFriedman

**Comment by srajama1** [comment]
@jhux2 @brian-kelley : Will any of the add issue or RefMaxwell changes cause this ?

**Comment by brian-kelley** [comment]
@srajama1 It's probably some interaction between my add changes and MueLu.

**Comment by srajama1** [comment]
Assuming this is a small problem, you could dump the matrix in Amesos2 and see what happened with it. It is simple 3 array CRS matrix.

**Comment by bathmatt** [comment]
Can we revert the offending change?  THis is hampering our L1 runs at this point as @cgcgcg had some fixes that improve performance.

**Comment by srajama1** [comment]
I thought you needed MM add for some other reason, no ?

**Comment by bathmatt** [comment]
Only to enable deprecated code removal.

**Comment by brian-kelley** [comment]
@srajama1 @bathmatt I'm making a PR to partially revert #6190. It fixed the deprecated-off issue where StaticProfile C didn't have enough entries allocated to hold the sum. I'm going to keep that change, but remove the fast path in Xpetra that uses the KokkosKernels-based add in Tpetra.

**Comment by bathmatt** [comment]
Thanks, I will test that to make sure it fixes our issue before pushing..

**Comment by brian-kelley** [comment]
@bathmatt Here is the branch: https://github.com/brian-kelley/Trilinos/tree/Partial6190Revert

All the Xpetra and MueLu tests (with and without deprecated on) passed in a local build.

**Comment by cgcgcg** [comment]
Are we sure that the change did not just expose a different issue? Cause it looks like everything is working out fine until we hit the coarse solve. We had similar error messages before when we tried to factor a singular system.

**Comment by srajama1** [comment]
@cgcgcg My first comment above :) ...

**Comment by cgcgcg** [comment]
@srajama1 Oh, right :-)

I had a look at one of the tests that fail, and switching the aggregation in MueLu from kokkos to host makes the test pass. I don't think that this is a Tpetra issue.

**Comment by srajama1** [comment]
Looks like none of us have a life :) .. Kokkos to host ?

**Comment by cgcgcg** [comment]
@srajama1 I switched from the kokkosified version of aggregation to the classical non-kokkos code.

**Comment by srajama1** [comment]
Ah, Why do we enable Kokkos on setup, that is not timed any way. Can we include a separate setup refactor flag and disable that, please ? @jhux2

**Comment by lucbv** [comment]
@srajama1 there is already a separate refactor flag and @jhux2 and I are aware of some of the aggregation issue and have decided that so far since setup is not a problem we will keep it on host and work on fixing some of the issues in aggregation in the meantime.
So overall nothing new... why did Empire decide to change the setup on GPU after we warned them though?

**Comment by cgcgcg** [comment]
@lucbv EMPIRE is running with "use kokkos refactor" = false, but we never put it in their default input deck, since their tests seemed to have run so far. Not sure how the combination of the new add and the new aggregation resulted in an issue.

**Comment by lucbv** [comment]
@cgcgcg the are a couple issues with the refactored aggregation currently and @jhux2 and I have been investigating it for some time.
In my opinion since there is a known bug it is a bit dangerous to leave the flag `use kokkos refactor=true`.
@jhux2 should probably weigh in since he has been spending more time on this than me and he has observed what triggers the issue with Empire.

**Comment by cgcgcg** [comment]
@rppawlo Could you change the three occurrences of "use kokkos refactor" in ElectromagneticSolverInterface.cpp to false, and see if that gets rid of the issue?

**Comment by rppawlo** [comment]
@cgcgcg - yes - I will try that, but does that turn off all use of kokkos in muelu?

**Comment by cgcgcg** [comment]
@rppawlo It does turn off kokkos for the setup phase, yes. (Linear algebra through Tpetra will still be on device.)

**Comment by srajama1** [comment]
@cgcgcg That flag confuses me as well, can we call it something with setup in it ?

**Comment by jhux2** [comment]
@bathmatt

**Comment by brian-kelley** [comment]
The add (6190) workaround has been merged, but I would still like to know if disabling kokkos refactor (but keeping the original add changes) also works. There are three possibilities:
- my add has a bug not caught by testing
- MueLu makes assumptions that non-fill complete matrices are globally indexed
- my add changes somehow expose an existing bug

**Comment by jhux2** [comment]
@rppawlo I've a separate report that Scot is seeing failures in the blob.  So I would like to request that `use kokkos refactor` be set to `false` in EMPIRE for the time being.

**Comment by rppawlo** [comment]
ok. I've got a build going now. Will push if everything passes...

**Comment by bathmatt** [comment]
@cgcgcg We are not sure.  HOwever, we need trilinos to solve our systems and this patch went from working to not working.. Could be something in the solver stack or in EMPIRE, not sure, think if we revert we can track it down.

**Comment by rppawlo** [comment]
Setting kokkos refactor to false did fix the issues in empire.

**Comment by rppawlo** [comment]
@cgcgcg - there are also two occurrences of "use kokkos refactor = true" in the Base class SolverInterface.cpp. Would you like those changed as well?

**Comment by cgcgcg** [comment]
@rppawlo Ok, so we should not revert the Tpetra changes. Also, once the autotester thinks it's ok, could you try the changes of PR #6268? On my workstation, this did fix the two failing EMPIRE tests you mentioned. I'm running the full EMPIRE test suite now.

**Comment by rppawlo** [comment]
Brian's partial revert was already pushed to develop about an hour ago. When #6268 goes in, do you want me to revert Brian's partial revert when testing? this is getting confusing :)

**Comment by cgcgcg** [comment]
@rppawlo That would be the idea, but let's wait on the autotester for #6268..

**Comment by rppawlo** [comment]
By the way - all of this testing is with the deprecated code path enabled. I have the full test suite running now. If that passes, I will restart with deprecated code off.

**Comment by rppawlo** [comment]
So the full set of empire tests are not passing with the refactor=false flag. Getting seg faults in a small number of tests in Refmaxwell.  This is not using #6268 yet. The OWave test in fluids is an example:
```
606: p=0 | CubeHexMesh: Building sub cells
606: [breve:230187] *** Process received signal ***
606: [breve:230187] Signal: Segmentation fault (11)
606: [breve:230187] Signal code: Address not mapped (1)
606: [breve:230187] Failing at address: (nil)
606: [breve:230187] [ 0] /lib64/libpthread.so.0(+0xf630)[0x7fc14c793630]
606: [breve:230187] [ 1] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libmuelu-adapters.so.12(_ZN5MueLu10RefMaxwellIdixN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEE16buildProlongatorEv+0x76b3)[0x7fc166aea243]
606: [breve:230187] [ 2] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libmuelu-adapters.so.12(_ZN5MueLu10RefMaxwellIdixN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEE7computeEb+0x4b56)[0x7fc166af3f56]
606: [breve:230187] [ 3] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libmuelu-adapters.so.12(_ZN5MueLu10RefMaxwellIdixN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEE11resetMatrixEN7Teuchos3RCPIN6Xpetra6MatrixIdixS6_EEEEb+0x9a)[0x7fc166affcda]
606: [breve:230187] [ 4] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libpanzer-stk.so.12(_ZN5MueLu10RefMaxwellIdixN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEEC1ERKN7Teuchos3RCPIN6Xpetra6MatrixIdixS6_EEEERNS8_13ParameterListEb+0xc0c)[0x7fc169aa008c]
606: [breve:230187] [ 5] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libpanzer-stk.so.12(_ZNK5Thyra36MueLuRefMaxwellPreconditionerFactoryIdixN6Kokkos6Compat23KokkosDeviceWrapperNodeINS1_6OpenMPENS1_9HostSpaceEEEE14initializePrecERKN7Teuchos3RCPIKNS_18LinearOpSourceBaseIdEEEEPNS_18PreconditionerBaseIdEENS_16ESupportSolveUseE+0x15b0)[0x7fc169ae4f80]
606: [breve:230187] [ 6] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libstratimikosbelos.so.12(_ZNK5Thyra29BelosLinearOpWithSolveFactoryIdE16initializeOpImplERKN7Teuchos3RCPIKNS_18LinearOpSourceBaseIdEEEES9_RKNS3_IKNS_18PreconditionerBaseIdEEEEbPNS_21LinearOpWithSolveBaseIdEENS_16ESupportSolveUseE+0x2459)[0x7fc1601570a9]
606: [breve:230187] [ 7] /home/rppawlo/install/roger-local-github-gnu-opt-openmp-shared/lib/libstratimikosbelos.so.12(_ZNK5Thyra29BelosLinearOpWithSolveFactoryIdE12initializeOpERKN7Teuchos3RCPIKNS_18LinearOpSourceBaseIdEEEEPNS_21LinearOpWithSolveBaseIdEENS_16ESupportSolveUseE+0x5b)[0x7fc160157e5b]
606: [breve:230187] [ 8] /home/rppawlo/empire2/src/circuit/libcircuit.so(_ZN5Thyra17linearOpWithSolveIdEEN7Teuchos3RCPINS_21LinearOpWithSolveBaseIT_EEEERKNS_28LinearOpWithSolveFactoryBaseIS4_EERKNS2_IKNS_12LinearOpBaseIS4_EEEENS_16ESupportSolveUseE+0xd0)[0x7fc16fbd29a0]
606: [breve:230187] [ 9] /home/rppawlo/empire2/src/em_solvers/libem_solvers.so(_ZNK6empire30ElectroMagneticSolverInterface25setupSchurComplementSolveERKN7Teuchos3RCPIN5Thyra12LinearOpBaseIdEEEERKNS2_IKNS3_28LinearOpWithSolveFactoryBaseIdEEEERKNS1_13ParameterListE+0xa97)[0x7fc16f1b2057]
606: [breve:230187] [10] /home/rppawlo/empire2/src/em_solvers/libem_solvers.so(_ZN6empire30ElectroMagneticSolverInterfaceC1E18MainParameterListsNS_13MeshContainerERNS_5utils9TimeStampEbN7Teuchos3RCPINS3_18MeshEvaluationBaseEEE+0x77b5)[0x7fc16f1c5525]
606: [breve:230187] [11] /home/rppawlo/empire2/EMPIRE-Fluid/src/empire-fluid.exe[0x422e49]
606: [breve:230187] [12] /home/rppawlo/empire2/EMPIRE-Fluid/src/empire-fluid.exe[0x421c26]
606: [breve:230187] [13] /lib64/libc.so.6(__libc_start_main+0xf5)[0x7fc14c3d8545]
606: [breve:230187] [14] /home/rppawlo/empire2/EMPIRE-Fluid/src/empire-fluid.exe[0x422140]
606: [breve:230187] *** End of error message ***
606: --------------------------------------------------------------------------
606: mpiexec noticed that process rank 0 with PID 230187 on node breve exited on signal 11 (Segmentation fault).
606: --------------------------------------------------------------------------
606: Failed to run step 0

```

**Comment by rppawlo** [comment]
Adding back the partial revert (and leaving in the kokkos refactor=false) still seg faults. usin refactor=true gets OWave passing again.

**Comment by cgcgcg** [comment]
@rppawlo I don't have EMPIRE-Fluid, are there any other tests that are failing this way?

**Comment by rppawlo** [comment]
EM and PIC tests are all passing. turns out it is only the OWave test problem failing in all of empire. Let me see if I can get you more info. It is calling the EM solve from the base repo so maybe just a mix of bad input parameters.

**Comment by rppawlo** [comment]
So the trilinos update to empire was successful last night. I wonder if pushing #6268 fixed the OWave test. I did not push the change to set kokkos refactor=false yet. @cgcgcg - do you still recommend we set the refactor to false? The test suite is clean at this point.

**Comment by cgcgcg** [comment]
@rppawlo I think that depends a bit on EMPIRE. The kokkos refactor code path is what MueLu is going to use in the future, and we do work a lot on it. But there could be bugs. It seems reasonable to me that EMPIRE would test with it, but we probably do not want to turn it on for production runs quite yet. (And for a single setup run the difference is probably negligible.)

**Comment by cgcgcg** [comment]
@rppawlo Can we try the revert of the partial revert with EMPIRE now?

**Comment by brian-kelley** [comment]
@rppawlo I'm emailing you a patch to undo the revert.

**Comment by brian-kelley** [comment]
@rppawlo I made PR #6277 which can be merged into develop once we know it doesn't break EMPIRE.

**Comment by rppawlo** [comment]
#6268 did not make it into the successful trilinos update from Wednesday night. It did make it into last night's round of testing. Last night the trilinos update failed in cuda builds only. Here's the output from the test: EMPIRE:Regression:ReadLinearSolverxmlPinningTest-MPI4
```
04:29:06 Transpose P (MueLu::TransPFactory)
04:29:06 matrixmatrix: kernel params -> 
04:29:06  [empty list]
04:29:06 
04:29:06 Computing Ac (MueLu::RAPFactory)
04:29:06 MxM: A x P
04:29:06 MxM: R x (AP) (explicit)
04:29:06 [cortado:226543] *** Process received signal ***
04:29:06 [cortado:226543] Signal: Floating point exception (8)
04:29:06 [cortado:226543] Signal code: Invalid floating point operation (7)
04:29:06 [cortado:226543] Failing at address: 0x3253394
```

**Comment by rppawlo** [comment]
This does not include the switch to kokkos refactor=false. I'm off for most of today but will test that asap.

**Comment by rppawlo** [comment]
Here's the stack trace for the cuda error in empire. It's in one single test only for the cuda build (EMPIRE:Regression:ReadLinearSolverxmlPinningTest-MPI4). Setting the kokkos refactor flag in the EMSolver does not fix this. This is against the current trilinos develop branch. @cgcgcg can you take a look at this test? 
```
(gdb) bt
#0  0x000000000324b064 in kk_get_suggested_vector_size (exec_space=KokkosKernels::Impl::Exec_CUDA, nnz=0, nr=0)
    at /home/rppawlo/Trilinos/packages/kokkos-kernels/src/common/KokkosKernels_ExecSpaceUtils.hpp:106
#1  get_suggested_vector_size (this=<optimized out>, this=<optimized out>, nnz=<optimized out>, nr=0)
    at /home/rppawlo/Trilinos/packages/kokkos-kernels/src/common/KokkosKernels_Handle.hpp:357
#2  KokkosSparse::Impl::KokkosSPGEMM<KokkosKernels::Experimental::KokkosKernelsHandle<unsigned int const, int const, double const, Kokkos::Cuda, Kokkos::CudaUVMSpace, Kokkos::CudaUVMSpace>, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<double*, Kokkos::CudaUVMSpace>, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<double*, Kokkos::CudaUVMSpace> >::getMaxRoughRowNNZ<Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> > > (
    this=this@entry=0x7ffdc0b7ce00, m=0, row_mapA_=..., entriesA_=..., row_pointers_begin_B=..., row_pointers_end_B=..., flops_per_row=0x7f39343ed080)
    at /home/rppawlo/Trilinos/packages/kokkos-kernels/src/sparse/impl/KokkosSparse_spgemm_impl_symbolic.hpp:2311
#3  0x0000000003266435 in KokkosSparse::Impl::KokkosSPGEMM<KokkosKernels::Experimental::KokkosKernelsHandle<unsigned int const, int const, double const, Kokkos::Cuda, Kokkos::CudaUVMSpace, Kokkos::CudaUVMSpace>, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<double*, Kokkos::CudaUVMSpace>, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<double*, Kokkos::CudaUVMSpace> >::KokkosSPGEMM_symbolic<Kokkos::View<unsigned int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> > > (this=this@entry=0x7ffdc0b7ce00, rowmapC_=...)
    at /home/rppawlo/Trilinos/packages/kokkos-kernels/src/sparse/impl/KokkosSparse_spgemm_impl_def.hpp:104
#4  0x000000000326759d in KokkosSparse::Impl::SPGEMM_SYMBOLIC<KokkosKernels::Experimental::KokkosKernelsHandle<unsigned int const, int const, double const, Kokkos::Cuda, Kokkos::CudaUVMSpace, Kokkos::CudaUVMSpace>, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, Kokkos::View<unsigned int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<1u> >, false, false>::spgemm_symbolic (handle=handle@entry=0x7ffdc0b7d080, m=m@entry=0, n=n@entry=0, k=k@entry=2, row_mapA=..., entriesA=..., transposeA=false, row_mapB=..., 
    entriesB=..., transposeB=false, row_mapC=...) at /home/rppawlo/Trilinos/packages/kokkos-kernels/src/sparse/impl/KokkosSparse_spgemm_symbolic_spec.hpp:212
#5  0x0000000003267c3a in KokkosSparse::Experimental::spgemm_symbolic<KokkosKernels::Experimental::KokkosKernelsHandle<unsigned int const, int const, double const, Kokkos::Cuda, Kokkos::CudaUVMSpace, Kokkos::CudaUVMSpace>, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<unsigned int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<unsigned int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::MemoryTraits<0u> > > (
    handle=handle@entry=0x7ffdc0b7d4b0, m=m@entry=0, n=n@entry=0, k=k@entry=2, row_mapA=..., entriesA=..., transposeA=false, row_mapB=..., entriesB=..., transposeB=false, 
    row_mapC=...) at /home/rppawlo/Trilinos/packages/kokkos-kernels/src/sparse/KokkosSparse_spgemm_symbolic.hpp:179
#6  0x0000000003273b3b in Tpetra::MMdetails::KernelWrappers<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace>, Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace> > >::mult_A_B_newmatrix_kernel_wrapper (Aview=..., Bview=..., Acol2Brow=..., Acol2Irow=..., 
    Bcol2Ccol=..., Icol2Ccol=..., C=..., Cimport=..., label=..., params=...) at /home/rppawlo/Trilinos/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_Cuda.hpp:212
#7  0x00000000032771a6 in Tpetra::MMdetails::mult_A_B_newmatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > (
    Aview=..., Bview=..., C=..., label=..., params=...) at /home/rppawlo/Trilinos/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:1704
#8  0x000000000327c1af in Tpetra::MatrixMatrix::Multiply<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > (A=..., 
    transposeA=<optimized out>, B=..., transposeB=<optimized out>, C=..., call_FillComplete_on_result=<optimized out>, label=..., params=...)
    at /home/rppawlo/Trilinos/packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp:260
#9  0x00000000012bb22c in Xpetra::MatrixMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::Multiply (A=..., 
    transposeA=<optimized out>, B=..., transposeB=<optimized out>, C=..., call_FillComplete_on_result=<optimized out>, doOptimizeStorage=<optimized out>, label=..., 
    params=...) at /home/rppawlo/Trilinos/packages/xpetra/sup/Utils/Xpetra_MatrixMatrix.hpp:361
#10 0x000000000165e684 in Xpetra::MatrixMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::Multiply (A=..., 
    transposeA=transposeA@entry=false, B=..., transposeB=transposeB@entry=false, C_in=..., fos=..., doFillComplete=true, doOptimizeStorage=true, label=..., params=...)
    at /home/rppawlo/Trilinos/packages/xpetra/sup/Utils/Xpetra_MatrixMatrix.hpp:444
#11 0x0000000001711827 in MueLu::RAPFactory<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::Build (this=0x265d4740, 
    fineLevel=..., coarseLevel=...) at /home/rppawlo/Trilinos/packages/muelu/src/Misc/MueLu_RAPFactory_def.hpp:210
#12 0x000000000131bfa2 in MueLu::TwoLevelFactoryBase::CallBuild (this=0x265d4740, requestedLevel=...)
    at /home/rppawlo/Trilinos/packages/muelu/src/Interface/../MueCentral/MueLu_TwoLevelFactoryBase.hpp:151
#13 0x0000000000a009ae in MueLu::Level::Get<Teuchos::RCP<Xpetra::Operator<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > > > (this=0x2660ce20, ename=..., factory=<optimized out>) at /home/rppawlo/Trilinos/packages/muelu/src/Interface/../MueCentral/MueLu_Level.hpp:204
#14 0x000000000184f0ea in MueLu::TopRAPFactory<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::Build (
---Type <return> to continue, or q <return> to quit---
    this=this@entry=0x7ffdc0b81460, coarseLevel=...) at /home/rppawlo/Trilinos/packages/muelu/src/MueCentral/MueLu_TopRAPFactory_def.hpp:76
#15 0x00000000016885bd in MueLu::Hierarchy<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::Setup (
    this=this@entry=0x2660bfd0, coarseLevelID=coarseLevelID@entry=1, fineLevelManager=..., coarseLevelManager=..., nextLevelManager=...)
    at /home/rppawlo/Trilinos/packages/muelu/src/MueCentral/MueLu_Hierarchy_def.hpp:419
#16 0x0000000000a27fd2 in MueLu::HierarchyManager<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::SetupHierarchy (
    this=0x26543530, H=...) at /home/rppawlo/Trilinos/packages/muelu/src/Interface/MueLu_HierarchyManager.hpp:241
#17 0x00000000009fb178 in MueLu::CreateXpetraPreconditioner<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> > (op=..., 
    inParamList=...) at /home/rppawlo/Trilinos/packages/muelu/adapters/xpetra/MueLu_CreateXpetraPreconditioner.hpp:97
#18 0x0000000000a201c8 in Thyra::MueLuPreconditionerFactory<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::initializePrec (this=<optimized out>, fwdOpSrc=..., prec=0x26587510, supportSolveUse=<optimized out>)
    at /home/rppawlo/Trilinos/packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_def.hpp:216
#19 0x000000000210e73c in Teko::PreconditionerInverseFactory::buildInverse (this=<optimized out>, linearOp=...)
    at /home/rppawlo/Trilinos/packages/teko/src/Teko_PreconditionerInverseFactory.cpp:180
#20 0x00000000020e77a1 in Teko::buildInverse (factory=..., A=...) at /home/rppawlo/Trilinos/packages/teko/src/Teko_InverseFactory.cpp:122
#21 0x00000000021195b1 in Teko::StratimikosFactory::initializePrec_Thyra (this=0x2656c070, fwdOpSrc=..., prec=<optimized out>)
    at /home/rppawlo/Trilinos/packages/teko/src/Teko_StratimikosFactory.cpp:210
#22 0x000000000211b9a5 in Teko::StratimikosFactory::initializePrec (this=0x2656c070, fwdOpSrc=..., prec=0x2655eed0, supportSolveUse=Thyra::SUPPORT_SOLVE_UNSPECIFIED)
    at /home/rppawlo/Trilinos/packages/teko/src/Teko_StratimikosFactory.cpp:139
#23 0x00000000022a6375 in Thyra::BelosLinearOpWithSolveFactory<double>::initializeOpImpl (this=this@entry=0x26557870, fwdOpSrc=..., approxFwdOpSrc=..., prec_in=..., 
    reusePrec=reusePrec@entry=false, Op=Op@entry=0x2655a470, supportSolveUse=<optimized out>)
    at /home/rppawlo/Trilinos/packages/stratimikos/adapters/belos/src/Thyra_BelosLinearOpWithSolveFactory_def.hpp:609
#24 0x00000000022a711b in Thyra::BelosLinearOpWithSolveFactory<double>::initializeOp (this=this@entry=0x26557870, fwdOpSrc=..., Op=Op@entry=0x2655a470, 
    supportSolveUse=supportSolveUse@entry=Thyra::SUPPORT_SOLVE_UNSPECIFIED)
    at /home/rppawlo/Trilinos/packages/stratimikos/adapters/belos/src/Thyra_BelosLinearOpWithSolveFactory_def.hpp:209
#25 0x00000000008d4b60 in initializeOp<double> (supportSolveUse=Thyra::SUPPORT_SOLVE_UNSPECIFIED, Op=..., fwdOp=..., lowsFactory=...)
    at /home/rppawlo/install/roger-local-github-cuda-opt-Pascal60-static/include/Thyra_LinearOpWithSolveFactoryHelpers.hpp:98
#26 Thyra::linearOpWithSolve<double> (lowsFactory=..., fwdOp=..., supportSolveUse=Thyra::SUPPORT_SOLVE_UNSPECIFIED)
    at /home/rppawlo/install/roger-local-github-cuda-opt-Pascal60-static/include/Thyra_LinearOpWithSolveFactoryHelpers.hpp:173
#27 0x00000000008f0d03 in empire::ElectroStaticSolverInterface::ElectroStaticSolverInterface (this=0x26382940, mpl=..., mesh_cont=..., time_stamp=..., mesh_eval=...)
    at /home/rppawlo/EMPIRE/src/em_solvers/ElectroStaticSolverInterface.cpp:211
#28 0x0000000000628aed in empire::em::meshSpecificMain<empire::MeshTraits<shards::Quadrilateral<4u>, 1> > (timer=..., comm=..., 
    start_timer=start_timer@entry=1574105967.128382, pl=..., mesh=..., time_stamp=...) at /home/rppawlo/EMPIRE/src/emMeshSpecificMain.hpp:109
#29 0x000000000046e5a0 in main (argc=<optimized out>, argv=<optimized out>) at /home/rppawlo/EMPIRE/src/main.cpp:194
(gdb) 
```

**Comment by brian-kelley** [comment]
@rppawlo I can see right away that the first input matrix to SPGEMM thinks it has 0 rows, which is why there's a divide by zero. The Tpetra matrix is getting into an invalid state somwhere.

**Comment by cgcgcg** [comment]
@rppawlo This is a bug in KokkosKernels. Fix is in https://github.com/kokkos/kokkos-kernels/pull/499

**Comment by rppawlo** [comment]
Thanks @cgcgcg ! Do we know if this will be snapshotted into trilinos ahead of a full kokkos update? Would @srajama1  be the one to do this?

**Comment by cgcgcg** [comment]
I'll let the KokkosKernels folks comment on that PR. We can then apply the same change to Trilinos directly.

**Comment by cgcgcg** [comment]
@rppawlo Should be fixed now in develop.

**Comment by rppawlo** [comment]
Thanks @cgcgcg !

**Comment by rppawlo** [comment]
So we were able to get a Trilinos update through over the weekend and were also able to disable all deprecated code in Trilinos. Thanks @brian-kelley and @cgcgcg ! If you would like me to test anything else, please let me know. We can close this ticket from empire perspective. I will close, but feel free to reopen if there is something else to clean up.
